src/java.desktop/macosx/native/libawt_lwawt/awt/CRobot.m

```diff
@@ -291,12 +303,25 @@ static inline void autoDelay(BOOL isMove) {
         CGEventSourceRef source = CGEventSourceCreate(kCGEventSourceStateHIDSystemState);
         CGKeyCode keyCode = GetCGKeyCode(javaKeyCode);
         CGEventRef event = CGEventCreateKeyboardEvent(source, keyCode, keyPressed);
+
         if (event != NULL) {
-            CGEventFlags flags = CGEventSourceFlagsState(kCGEventSourceStateHIDSystemState);
-            if ((flags & kCGEventFlagMaskSecondaryFn) != 0) {
-                flags ^= kCGEventFlagMaskSecondaryFn;
-                CGEventSetFlags(event, flags);
+            int flagMaskValue = GetCGKeyMask(keyCode);
+            if (OSX_Undefined != flagMaskValue) {
+                if (keyCode == OSX_CapsLock) {
+                    if (keyPressed) {
+                     initFlags ^= flagMaskValue;
+                    }
+                } else {
+                    initFlags = keyPressed
+                                ? (initFlags | flagMaskValue)    // add flag bits if modifier key pressed
+                                : (initFlags & ~flagMaskValue);  // clear flag bits if modifier key released
+                }
             }
+
+            CGEventFlags flags = CGEventSourceFlagsState(kCGEventSourceStateHIDSystemState);
+            flags  = (initFlags & allModifiersMask) | (flags & (!allModifiersMask));
```

Comment:
Should `!allModifiersMask` be `~allModifiersMask`? That is _bitwise_ NOT instead of _logical_ NOT.

```suggestion
            flags = (initFlags & allModifiersMask) | (flags & (!allModifiersMask));
```

Reply:
That is correct, Thanks for catching it. It should be `~`.